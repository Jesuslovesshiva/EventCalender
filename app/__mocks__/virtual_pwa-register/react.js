import {useState} from 'react'

const useRegisterSW = jest.fn(() => {
  const [needRefresh, setNeedRefresh] = useState(true)

  return {
    needRefresh: [needRefresh, setNeedRefresh],
    updateServiceWorker: () => {}
  }
})

// eslint-disable-next-line import/prefer-default-export
export {useRegisterSW}
