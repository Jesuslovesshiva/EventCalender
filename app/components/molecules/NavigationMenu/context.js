import React, {useContext} from 'react'

const NavigationMenuContext = React.createContext()

const useNavigationMenuContext = () => useContext(NavigationMenuContext)

export {useNavigationMenuContext}
export default NavigationMenuContext
