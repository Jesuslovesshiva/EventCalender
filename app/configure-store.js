import {
  combineReducers,
  configureStore as configureToolkitStore,
  isPlain
} from '@reduxjs/toolkit'
import {createBrowserHistory} from '@tanstack/react-location'
import {createReduxHistoryContext} from 'redux-first-history'
import {createInjectorsEnhancer} from 'redux-injectors'
import createSagaMiddleware from 'redux-saga'

import ActionSubscription, {
  actionListenerMiddleware
} from '@/components/organisms/ActionSubscription'

const actionEmitter = new ActionSubscription.ActionEmitter()

const {createReduxHistory, routerMiddleware, routerReducer} =
  createReduxHistoryContext({
    history: createBrowserHistory()
  })

const sagaMiddleware = createSagaMiddleware()

const configureStore = () => {
  const {run: runSaga} = sagaMiddleware
  const middlewares = [
    sagaMiddleware,
    routerMiddleware,
    actionListenerMiddleware(actionEmitter)
  ]

  const createReducer = injectedReducers =>
    combineReducers({
      router: routerReducer,
      ...injectedReducers
    })
  const enhancers = [createInjectorsEnhancer({createReducer, runSaga})]

  const isSerializable = value =>
    value instanceof Error ||
    value instanceof File ||
    value instanceof Response ||
    isPlain(value)

  const store = configureToolkitStore({
    // eslint-disable-next-line no-undef
    devTools: process.env.NODE_ENV !== 'production',
    enhancers,
    middleware: getDefaultMiddleware => [
      ...getDefaultMiddleware({serializableCheck: {isSerializable}}),
      ...middlewares
    ],
    reducer: createReducer()
  })

  store.injectedReducers = {}
  store.injectedSagas = {}

  return store
}

const store = configureStore()
const history = createReduxHistory(store)

export {configureStore, history, actionEmitter}

export default store
