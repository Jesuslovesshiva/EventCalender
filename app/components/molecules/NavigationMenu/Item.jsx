import {MenuItem} from '@szhsin/react-menu'
import {Link} from '@tanstack/react-location'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import {useRef} from 'react'

// prettier-ignore
const getItemClassName =
  ({extraClasses, textColor = 'text-primary'} = {}) =>
    ({hover}) =>
      classNames(
        'font-bold text-[11px] leading-[13px] rounded',
        'transition-colors motion-reduce:transition-none',
        'focus:bg-red',
        {
          'bg-primary text-white': hover,
          [textColor]: !hover
        },
        extraClasses
      )

const NavigationMenuItem = ({children, isDisabled, to, ...props}) => {
  const linkRef = useRef(null)

  const onKeyUp = event => {
    if (event.code === 'Enter' && !isDisabled) {
      linkRef.current.click?.()
    }
  }

  return (
    <MenuItem
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      className={getItemClassName()}
      onKeyUp={onKeyUp}
    >
      <Link
        className="block px-1.5 py-[3px]"
        disabled={isDisabled}
        to={to}
        _ref={linkRef}
      >
        {children}
      </Link>
    </MenuItem>
  )
}

NavigationMenuItem.propTypes = {
  children: PropTypes.node,
  isDisabled: PropTypes.bool,
  to: PropTypes.string
}

NavigationMenuItem.defaultProps = {
  children: null,
  isDisabled: false,
  to: null
}

export {getItemClassName}
export default NavigationMenuItem
