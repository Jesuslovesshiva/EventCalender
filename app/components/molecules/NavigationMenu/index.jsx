import {ControlledMenu, useMenuState} from '@szhsin/react-menu'
import classNames from 'classnames'
import noop from 'lodash/noop'
import PropTypes from 'prop-types'
import {useRef} from 'react'

import animation from '@/utils/menu-animation.module.css'

import NavigationMenuContext from './context'

const getMenuClassName =
  extraClasses =>
    ({state}) =>
      classNames(
        'shadow bg-white px-1.5 py-2.5 border-0 rounded-bl min-w-[134px] space-y-1',
        {'pointer-events-none': ['closing', 'closed'].includes(state)},
        animation.slideReverse,
        extraClasses
      )

const NavigationMenu = props => {
  const buttonRef = useRef(null)
  const menuRef = useRef(null)
  const openByKeyboard = useRef(false)
  const [menuProps, toggleMenu] = useMenuState({transition: true})

  const openMenu = () => toggleMenu(true)

  const closeMenu = () => {
    toggleMenu(false)

    if (openByKeyboard.current) {
      buttonRef.current?.focus?.()
    }

    openByKeyboard.current = false
  }

  const onButtonMouseLeave = event => {
    if (
      !event.relatedTarget?.nodeType ||
      !menuRef.current?.contains(event.relatedTarget)
    ) {
      closeMenu()
    }
  }

  const onMenuMouseLeave = event => {
    if (
      !event.relatedTarget?.nodeType ||
      !buttonRef.current?.contains(event.relatedTarget)
    ) {
      closeMenu()
    }
  }

  const onButtonKeyDown = event => {
    if (['ArrowDown', 'ArrowUp'].includes(event.code)) {
      openByKeyboard.current = true
      openMenu()
    }
  }

  return (
    <>
      {props.menuButton({
        isOpen: ['opening', 'open'].includes(menuProps.state),
        onKeyDown: onButtonKeyDown,
        onMouseEnter: openMenu,
        onMouseLeave: onButtonMouseLeave,
        ref: buttonRef
      })}

      <ControlledMenu
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...menuProps}
        menuClassName={getMenuClassName()}
        ref={menuRef}
        anchorRef={buttonRef}
        onMouseLeave={onMenuMouseLeave}
        onClose={closeMenu}
        submenuCloseDelay={0}
        submenuOpenDelay={0}
        portal
        transition
      >
        <NavigationMenuContext.Provider value={{closeMenu}}>
          {props.children}
        </NavigationMenuContext.Provider>
      </ControlledMenu>
    </>
  )
}

NavigationMenu.propTypes = {
  children: PropTypes.node,
  menuButton: PropTypes.func
}

NavigationMenu.defaultProps = {
  children: null,
  menuButton: noop
}

export {getMenuClassName}
export default NavigationMenu
