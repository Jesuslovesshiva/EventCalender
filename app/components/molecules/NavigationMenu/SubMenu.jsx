import {SubMenu} from '@szhsin/react-menu'
import {Link} from '@tanstack/react-location'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import {useMemo, useRef} from 'react'

import ArrowDropDownIcon from '@/components/atoms/Icon/ArrowDropDown'

import {getItemClassName} from './Item'
import {useNavigationMenuContext} from './context'

import {getMenuClassName} from './index'

const NavigationSubMenu = ({isLabelDisabled, label, to, ...props}) => {
  const linkRef = useRef(null)
  const subMenuRef = useRef(null)
  const {closeMenu} = useNavigationMenuContext()

  const labelTextColor = useMemo(() => {
    if (isLabelDisabled) {
      return 'text-gray-dark'
    }
    return 'text-primary'
  }, [isLabelDisabled])

  const onKeyUp = event => {
    if (event.code === 'Enter' && !isLabelDisabled) {
      linkRef.current.click?.()
      closeMenu()
    }

    if (['ArrowRight', 'Space'].includes(event.code)) {
      subMenuRef.current.openMenu?.('first')
    }
  }

  const onClick = () => {
    closeMenu()
  }

  const onMouseEnter = () => {
    subMenuRef.current.openMenu?.()
  }

  return (
    <SubMenu
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      instanceRef={subMenuRef}
      menuClassName={getMenuClassName('rounded-r')}
      itemProps={{
        className: getItemClassName({
          extraClasses: {'cursor-default': isLabelDisabled},
          textColor: labelTextColor
        }),
        onClick,
        onKeyUp,
        onMouseEnter
      }}
      offsetX={6}
      offsetY={-10}
      openTrigger="none"
      label={
        <Link
          className={classNames('flex place-content-between px-1.5 py-[3px]')}
          disabled={isLabelDisabled}
          tabIndex="-1"
          to={to}
          _ref={linkRef}
        >
          {label}
          <ArrowDropDownIcon className="text-lg -rotate-90 ml-1" />
        </Link>
      }
    />
  )
}

NavigationSubMenu.propTypes = {
  isLabelDisabled: PropTypes.bool,
  label: PropTypes.node,
  to: PropTypes.string
}

NavigationSubMenu.defaultProps = {
  isLabelDisabled: false,
  label: null,
  to: null
}

export default NavigationSubMenu
