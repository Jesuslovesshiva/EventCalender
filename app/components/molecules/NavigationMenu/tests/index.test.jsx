import {act, fireEvent, waitFor} from '@testing-library/react'
import {omit} from 'lodash'

import {history} from '@/configure-store'
import render from '@/utils/test-utils/custom-render'

import NavigationMenuItem from '../Item'
import NavigationMenu from '../index'

const defaultProps = {
  children: <NavigationMenuItem>navigationMenuItem</NavigationMenuItem>,
  menuButton: props => <button {...omit(props, 'isOpen')}>menuButton</button>
}

const options = {wrapperProps: {hasRouter: true}}
const renderComponent = props =>
  render(<NavigationMenu {...defaultProps} {...props} />, options)

describe('NavigationMenu', () => {
  it('should match the snapshot', () => {
    const {container} = renderComponent()
    fireEvent.mouseEnter(container.firstChild)
    expect(document.body).toMatchSnapshot()
  })

  it('should open on arrow down key press', () => {
    const {getByText, queryByText} = renderComponent()
    const button = getByText('menuButton')

    expect(queryByText('navigationMenuItem')).not.toBeInTheDocument()
    fireEvent.keyDown(button, {code: 'ArrowDown'})
    expect(queryByText('navigationMenuItem')).toBeInTheDocument()
  })

  it('should not open on another key press', () => {
    const {getByText, queryByText} = renderComponent()
    const button = getByText('menuButton')

    fireEvent.keyDown(button, {code: 'Enter'})
    expect(queryByText('navigationMenuItem')).not.toBeInTheDocument()
  })

  it('should close menu on button mouse leave', () => {
    const {getByText, getByRole} = renderComponent()
    const button = getByText('menuButton')

    fireEvent.mouseEnter(button)
    expect(getByRole('menu')).toBeInTheDocument()
    fireEvent.mouseLeave(button)
    expect(getByRole('menu')).toHaveClass('szh-menu--state-closing')
  })

  it('should not close menu on button mouse leave when hovering the menu', () => {
    const {getByText, getByRole} = renderComponent()
    const button = getByText('menuButton')

    fireEvent.mouseEnter(button)
    expect(getByRole('menu')).toBeInTheDocument()
    fireEvent.mouseLeave(button, {
      relatedTarget: getByText('navigationMenuItem')
    })
    expect(getByRole('menu')).not.toHaveClass('szh-menu--state-closing')
  })

  it('should close menu on menu mouse leave', () => {
    const {getByText, getByRole} = renderComponent()
    const button = getByText('menuButton')

    fireEvent.mouseEnter(button)
    const menu = getByRole('menu')

    expect(menu).toBeInTheDocument()
    fireEvent.mouseLeave(menu)
    expect(menu).toHaveClass('szh-menu--state-closing')
  })

  it('should not close menu on menu mouse leave when hovering the button', () => {
    const {getByText, getByRole} = renderComponent()
    const button = getByText('menuButton')

    fireEvent.mouseEnter(button)
    const menu = getByRole('menu')

    expect(menu).toBeInTheDocument()
    fireEvent.mouseLeave(menu, {relatedTarget: button})
    expect(menu).not.toHaveClass('szh-menu--state-closing')
  })

  it('should focus menu button on close if opened via keyboard', () => {
    const {getByText, getByRole} = renderComponent()
    const button = getByText('menuButton')

    fireEvent.keyDown(button, {code: 'ArrowDown'})
    const menu = getByRole('menu')

    fireEvent.keyDown(menu, {code: 'Escape', key: 'Escape'})
    expect(menu).toHaveClass('szh-menu--state-closing')
    expect(document.activeElement).toBe(button)
  })

  describe('Item', () => {
    it('should trigger Link click on enter key up', async () => {
      const linkPath = '/foo'
      const {getByText, getByRole} = renderComponent({
        children: (
          <NavigationMenuItem to={linkPath}>
            navigationMenuItem
          </NavigationMenuItem>
        )
      })
      fireEvent.mouseEnter(getByText('menuButton'))
      const menuItem = getByRole('menuitem')

      fireEvent.keyUp(menuItem, {code: 'Enter'})
      await waitFor(() => expect(history.location.pathname).toBe(linkPath))
      act(() => {
        history.push('/')
      })
    })

    it('should not trigger Link click on disabled item', () => {
      const {getByText, getByRole} = renderComponent({
        children: (
          <NavigationMenuItem isDisabled>navigationMenuItem</NavigationMenuItem>
        )
      })
      fireEvent.mouseEnter(getByText('menuButton'))
      const menuItem = getByRole('menuitem')
      const link = getByText('navigationMenuItem')
      link.onclick = jest.fn()

      fireEvent.keyUp(menuItem, {code: 'Enter'})
      expect(link.onclick).not.toHaveBeenCalled()
    })

    it('should disable link', () => {
      const {getByText} = renderComponent({
        children: (
          <NavigationMenuItem to="/foo" isDisabled>
            navigationMenuItem
          </NavigationMenuItem>
        )
      })
      fireEvent.mouseEnter(getByText('menuButton'))
      const link = getByText('navigationMenuItem')
      expect(link).not.toHaveAttribute('href')
      expect(link).toHaveAttribute('aria-disabled', 'true')
    })
  })
})
