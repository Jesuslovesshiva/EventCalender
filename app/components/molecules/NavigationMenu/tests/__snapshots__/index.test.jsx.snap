// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationMenu should match the snapshot 1`] = `
<body>
  <div>
    <button>
      menuButton
    </button>
  </div>
  <div
    class="szh-menu-container szh-menu-container--itemTransition"
    style="position: relative;"
  >
    <ul
      aria-label="Menu"
      class="szh-menu szh-menu--state-opening szh-menu--dir-bottom shadow bg-white px-1.5 py-2.5 border-0 rounded-bl min-w-[134px] space-y-1 slideReverse"
      role="menu"
      style="margin: 0px; position: absolute; left: 0px; top: 0px;"
      tabindex="-1"
    >
      <div
        style="position: absolute; left: 0px; top: 0px;"
        tabindex="-1"
      />
      <li
        class="szh-menu__item font-bold text-[11px] leading-[13px] rounded transition-colors motion-reduce:transition-none focus:bg-red text-primary"
        role="menuitem"
        tabindex="-1"
      >
        <a
          class="block px-1.5 py-[3px] active"
          href="/"
        >
          navigationMenuItem
        </a>
      </li>
    </ul>
  </div>
</body>
`;
